### PR TITLE
Split vcmiclient in two

### PR DESCRIPTION
**Pull request by smanolloff**
Similar to vcmiserver (app) and vcmiservercommon (lib), now there is vcmiclient (app) and vcmiclientcommon (lib).

This will allow to compile other clients (such as an ML client) which can now link against `vcmiclientcommon` to avoid the need to duplicate all of the code related to client initialization, networking, etc.

(I have not tried to cross-compile it, so I will rely on the CI results for iOS, Android and Windows builds)

### CODE REVIEW

File: clientapp/EntryPoint.cpp
Comment (IvanSavenko) [inline clientapp/EntryPoint.cpp:1]: A bit weird to have app startup in this file in clientapp project, but app shutdown - in CMT in client project.

I understand why it ended up that way, but still weird. Need to think how to make it nicer.

File: client/CMakeLists.txt [hunk 1]
```diff
@@ -396,76 +396,58 @@ set(client_HEADERS
...
 	)
 endif()
 
-assign_source_group(${client_SRCS} ${client_HEADERS} VCMI_client.rc)
+assign_source_group(${vcmiclientcommon_SRCS} ${vcmiclientcommon_HEADERS})
 
 if(ANDROID)
-	add_library(vcmiclient SHARED ${client_SRCS} ${client_HEADERS})
-	set_target_properties(vcmiclient PROPERTIES
+	add_library(vcmiclientcommon SHARED ${vcmiclientcommon_SRCS} ${vcmiclientcommon_HEADERS})
```
Comment (IvanSavenko) [inline client/CMakeLists.txt:410]: How `clientapp/EntryPoint.cpp` is included on Android now if it is no longer part of vcmiclient shared library?

**Review by IvanSavenko — COMMENTED** [review]
Considering changes - might be better to double-check that iOS and Android artifacts still work

**Comment by smanolloff** [comment]
> Considering changes - might be better to double-check that iOS and Android artifacts still work

I will try to set up build environments for ios/android locally and check.

**Comment by kambala-decapitator** [comment]
> > Considering changes - might be better to double-check that iOS and Android artifacts still work
> 
> I will try to set up build environments for ios/android locally and check.

it's much easier to grab artifacts from CI and install on your device(s). I have both iOS and Android if you need help in testing.

**Comment by smanolloff** [comment]
> > > Considering changes - might be better to double-check that iOS and Android artifacts still work
> > 
> > 
> > I will try to set up build environments for ios/android locally and check.
> 
> it's much easier to grab artifacts from CI and install on your device(s). I have both iOS and Android if you need help in testing.

That's really kind of you, thanks. I may ask you to try later, but for now I want to make the runs complete successfully -- they did once, but are constantly being cancelled now, do you know how can I restart the builds without a git push?

**Comment by IvanSavenko** [comment]
Generally, you need a git push.
You can make a fake one:
`git commit --amend`
`git push -f`
Empty amend would still change commit date, and as result - hash.

Repository owners (e.g. me) can also restart CI without push. Let me do that...

File: client/CMakeLists.txt (hunk 1 above)
Comment (smanolloff) [reply client/CMakeLists.txt]: correct - it was not. I have changed clientapp to be built as a shared library if ANDROID is set

**Comment by kambala-decapitator** [comment]
> but are constantly being cancelled now

if they're cancelled because one of the builds fails, you can disable that by setting [this variable](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast) to false

**Comment by smanolloff** [comment]
fixed & verified Android artifact as well -- it works now \o/

File: client/CMT.h [hunk 2]
```diff
@@ -20,8 +20,6 @@ extern SDL_Surface *screen;      // main screen surface
 extern SDL_Surface *screen2;     // and hlp surface (used to store not-active interfaces layer)
 extern SDL_Surface *screenBuf; // points to screen (if only advmapint is present) or screen2 (else) - should be used when updating controls which are not regularly redrawed
 
+// defined in clientapp EntryPoint
 void handleQuit(bool ask = true);
-
-/// Notify user about encountered fatal error and terminate the game
-/// TODO: decide on better location for this method
-[[noreturn]] void handleFatalError(const std::string & message, bool terminate);
+void handleFatalError(const std::string & message, bool terminate);
```
Comment (IvanSavenko) [inline client/CMT.h:25]: Why was comment & `[[noreturn]]` attribute removed?

File: clientapp/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+      PROJECT_LABEL "VCMI_client"
+  )
+  set_property(DIRECTORY ${CMAKE_SOURCE_DIR} PROPERTY VS_STARTUP_PROJECT vcmiclient)
+  if(NOT ENABLE_DEBUG_CONSOLE)
+    set_target_properties(vcmiclient PROPERTIES WIN32_EXECUTABLE)
+    target_link_libraries(vcmiclient SDL2::SDL2main)
+  endif()
+  target_compile_definitions(vcmiclient PRIVATE WINDOWS_IGNORE_PACKING_MISMATCH)
+
+# TODO: very hacky, find proper solution to copy AI dlls into bin dir
```
Comment (kambala-decapitator) [inline clientapp/CMakeLists.txt:59]: please fix indentation

**Review by kambala-decapitator — COMMENTED** [review]
please change indentation in the new CML file to use tabs instead of spaces

in general looks fine to me

**Comment by smanolloff** [comment]
> please change indentation in the new CML file to use tabs instead of spaces
> 
> in general looks fine to me

👍 done

**Review by kambala-decapitator — APPROVED** [review]

**Comment by pppbb** [comment]
After this commit linux version don't build anymore.

**Comment by IvanSavenko** [comment]
Not related to this PR, there were issues with boost 1.86. Can't find issue about it for some reason though...

**Comment by Gigas002** [comment]
@pppbb Could you try out removing the build caches and rebuilding the package, or rebuilding in a clean chroot? The package builds fine for me without old caches

**Comment by AlexFolland** [comment]
Deleting the issue ticket must have been a mistake, since it doesn't seem like it could be intended.  Can github staff restore it?

**Comment by Crandel** [comment]
I've created a new issue #4561

**Pull request merged**